multiapi-engine/src/main/java/com/sngular/api/generator/plugin/asyncapi/AsyncApiGenerator.java

```diff
@@ -167,29 +166,30 @@ private void checkRequiredOrCombinatorExists(final SchemaObject schema, final bo
 
   private Map<String, JsonNode> getAllSchemas(final Path ymlParentPath, final JsonNode node) {
     final Map<String, JsonNode> totalSchemas = new HashMap<>();
-    final List<JsonNode> referenceList = node.findValues("$ref");
+    final List<JsonNode> referenceList = node.findValues(REF);
     referenceList.forEach(reference ->
-        processReference(node, reference, ymlParentPath, totalSchemas)
+        processReference(node, reference, ymlParentPath, totalSchemas, referenceList)
     );
     final List<JsonNode> messagesList = node.findValues("messages");
     final List<JsonNode> schemasList = node.findValues("schemas");
-    schemasList.forEach(schema -> schema.fields().forEachRemaining(fieldSchema -> totalSchemas.putIfAbsent(fieldSchema.getKey().toUpperCase(), fieldSchema.getValue())));
+    schemasList.forEach(schema -> schema.fields().forEachRemaining(fieldSchema -> totalSchemas.putIfAbsent(("schemas/" + fieldSchema.getKey()).toUpperCase(),
+                                                                   fieldSchema.getValue())));
     messagesList.forEach(message ->
                            message.fields().forEachRemaining(fieldSchema -> {
-                             if (fieldSchema.getValue().has("payload")) {
-                               final var payload = fieldSchema.getValue().get("payload");
-                               if (!payload.has("$ref")) {
-                                 totalSchemas.put(fieldSchema.getKey().toUpperCase(), payload);
+                             if (fieldSchema.getValue().has(PAYLOAD)) {
+                               final var payload = fieldSchema.getValue().get(PAYLOAD);
+                               if (!payload.has(REF)) {
+                                 totalSchemas.put(("messages/" + fieldSchema.getKey()).toUpperCase(), payload);
```

Comment:
I think you should add the part of the URI as constant to follow the other changes realized, try to separate the content and the slash in two different constants